File: game.py
import pygame
from entities import Player, Background, Bullet, Bomb, Enemy, WaveCannon, EnemyBullet, ShieldCapsule, WeaponCapsule, LaserCapsule, LaserBullet, Boss
from audio import audio_manager
import random
import math

class Game:
    def __init__(self, width, height, joystick=None, initial_score=0):
        self.width = width
        self.height = height
        self.joystick = joystick
        self.player = Player(self.width / 2, self.height - 100)
        self.background = Background(self.width, self.height)
        self.keys = {}
        
        self.bullets = []
        self.enemy_bullets = []
        self.bombs = []
        self.enemies = []
        self.items = []
        
        self.score = initial_score
        self.lives = 3
        self.game_over = False
        self.boss_active = False
        self.boss_spawned = False
        self.stage_clear = False
        
        self.enemy_spawn_timer = 0
        self.enemy_spawn_interval = 60
        
        self.ground_spawn_timer = 0
        self.ground_spawn_interval = 180

        self.frame_count = 0
        self.phaeon_spawn_timer = 0
        self.phaeon_spawn_interval = 300

    def trigger_rumble(self, low, high, duration):
        if self.joystick and hasattr(self.joystick, 'rumble'):
            try:
                self.joystick.rumble(low, high, duration)
            except Exception as e:
                pass

    def update(self):
        if self.game_over or self.stage_clear:
            return

        if self.keys.get(pygame.K_t):
            if self.frame_count < 10500:
                self.frame_count = 10500

        self.frame_count += 1

        if self.frame_count >= 10800 and not self.boss_active and not self.boss_spawned:
            self.enemies.append(Boss(self.width / 2, -50))
            self.boss_active = True
            self.boss_spawned = True
            audio_manager.play_boss_bgm()

        if self.boss_active:
            bosses = [e for e in self.enemies if e.type == 'boss']
            if not bosses:
                self.boss_active = False
                self.stage_clear = True
                self.enemies.clear()
                self.enemy_bullets.clear()

        self.background.update()
        self.player.update(self.keys, self.width, self.height)

        if getattr(self.player, 'rumble_trigger', None):
            trigger = self.player.rumble_trigger
            self.player.rumble_trigger = None
            if trigger == 'charge_complete':
                self.trigger_rumble(0.0, 0.6, 120)
            elif trigger == 'charge_complete2':
                self.trigger_rumble(0.0, 1.0, 200)

        if getattr(self.player, 'shoot_normal', False):
            self.player.shoot_normal = False
            if getattr(self.player, 'has_laser', False):
                self.bullets.append(LaserBullet(self.player.x, self.player.y - self.player.height / 2))
                self.player.cooldown_air = 12
            else:
                weapon_level = getattr(self.player, 'weapon_level', 1)
                if weapon_level >= 3:
                    angles = [-25, -8, 8, 25]
                    offsets = [-15, -5, 5, 15]
                    for angle, offset in zip(angles, offsets):
                        rad = math.radians(angle)
                        vx = 10 * math.sin(rad)
                        vy = -10 * math.cos(rad)
                        self.bullets.append(Bullet(self.player.x + offset, self.player.y - self.player.height / 2, vx, vy))
                elif weapon_level == 2:
                    self.bullets.append(Bullet(self.player.x - 10, self.player.y - self.player.height / 2))
                    self.bullets.append(Bullet(self.player.x + 10, self.player.y - self.player.height / 2))
                else:
                    self.bullets.append(Bullet(self.player.x, self.player.y - self.player.height / 2))
                self.player.cooldown_air = 10
            audio_manager.play('laser')
            
        if getattr(self.player, 'shoot_wave', False):
            self.player.shoot_wave = False
            self.bullets.append(WaveCannon(self.player.x, self.player.y - self.player.height / 2))
            audio_manager.play('wave_cannon')
            self.trigger_rumble(0.6, 0.4, 250)
            
        if getattr(self.player, 'shoot_diffusion_wave', False):
            self.player.shoot_diffusion_wave = False
            angles = [-30, -15, 0, 15, 30]
            base_speed = 12
            for angle in angles:
                rad = math.radians(angle)
                vx = base_speed * math.sin(rad)
                vy = -base_speed * math.cos(rad)
                self.bullets.append(WaveCannon(self.player.x, self.player.y - self.player.height / 2, vx=vx, vy=vy))
            audio_manager.play('wave_cannon')
            self.trigger_rumble(0.9, 0.7, 450)
                
        if self.keys.get(pygame.K_x) or self.keys.get(pygame.K_c):
            if self.player.cooldown_ground <= 0:
                target_x = self.player.x
                target_y = self.player.y - self.player.reticle_distance
                self.bombs.append(Bomb(self.player.x, self.player.y, target_x, target_y))
                self.player.cooldown_ground = 60
                audio_manager.play('bomb_launch')

        if not self.boss_spawned:
            self.enemy_spawn_timer += 1
            if self.enemy_spawn_timer > self.enemy_spawn_interval:
                self.spawn_enemy('air')
                self.enemy_spawn_timer = 0
                if self.enemy_spawn_interval > 20:
                    self.enemy_spawn_interval -= 0.5
                    
            self.ground_spawn_timer += 1
            if self.ground_spawn_timer > self.ground_spawn_interval:
                self.spawn_enemy('ground')
                self.ground_spawn_timer = 0

            # Phaeon spawning logic after 60 seconds (3600 frames)
            if self.frame_count > 3600:
                self.phaeon_spawn_timer += 1
                if self.phaeon_spawn_timer > self.phaeon_spawn_interval:
                    self.phaeon_spawn_timer = 0
                    num_phaeons = random.randint(1, 3)
                    base_x = random.uniform(80, self.width - 80)
                    for i in range(num_phaeons):
                        spawn_x = max(30, min(self.width - 30, base_x + (i - (num_phaeons - 1) / 2) * 40))
                        spawn_y = -50 - i * 30
                        self.enemies.append(Enemy(spawn_x, spawn_y, 'phaeon'))

        for b in self.bullets: b.update()
        for b in self.bombs:
            was_exploded = b.exploded
            b.update()
            if b.exploded and not was_exploded:
                self.trigger_rumble(0.4, 0.2, 150)
        for e in self.enemies:
            e.update()
            if getattr(e, 'shoot_now', False):
                e.shoot_now = False
                if e.type == 'phaeon':
                    dx = self.player.x - e.x
                    dy = self.player.y - e.y
                    dist = math.hypot(dx, dy)
                    if dist > 0:
                        bullet_speed = 6.0
                        vx = (dx / dist) * bullet_speed
                        vy = (dy / dist) * bullet_speed
                    else:
                        vx = 0.0
                        vy = 6.0
                    self.enemy_bullets.append(EnemyBullet(e.x, e.y, vx, vy))
                elif e.type == 'boss':
                    angles = [-40, -20, 0, 20, 40]
                    bullet_speed = 5.0
                    for angle in angles:
                        rad = math.radians(angle)
                        vx = bullet_speed * math.sin(rad)
                        vy = bullet_speed * math.cos(rad)
                        self.enemy_bullets.append(EnemyBullet(e.x, e.y + 20, vx, vy))
                else:
                    self.enemy_bullets.append(EnemyBullet(e.x, e.y + e.height / 2))
        for eb in self.enemy_bullets: eb.update()
        for item in self.items: item.update()

        self.bullets = [b for b in self.bullets if not b.marked_for_deletion]
        self.enemy_bullets = [eb for eb in self.enemy_bullets if not eb.marked_for_deletion]
        self.bombs = [b for b in self.bombs if not b.marked_for_deletion]
        self.enemies = [e for e in self.enemies if not e.marked_for_deletion]
        self.items = [i for i in self.items if not i.marked_for_deletion]

        self.check_collisions()

    def draw(self, surface):
        self.background.draw(surface)
        
        for e in self.enemies:
            if e.type == 'ground': e.draw(surface)
            
        for b in self.bombs: b.draw(surface)
        for item in self.items: item.draw(surface)
        
        if not self.game_over:
            self.player.draw(surface)
            
        for e in self.enemies:
            if e.type in ('air', 'phaeon', 'boss'): e.draw(surface)
            
        for b in self.bullets: b.draw(surface)
        for eb in self.enemy_bullets: eb.draw(surface)

    def spawn_enemy(self, type_):
        x = random.uniform(30, self.width - 30)
        y = -50
        self.enemies.append(Enemy(x, y, type_))

    def check_collisions(self):
        for bullet in self.bullets:
            for enemy in [e for e in self.enemies if e.type in ('air', 'phaeon', 'boss')]:
                if self.is_colliding(bullet, enemy):
                    if enemy.type == 'boss':
                        if enemy.state in ('ENTER', 'HOVER'):
                            enemy.hp -= 1
                            enemy.flash_timer = 5
                            if not isinstance(bullet, (WaveCannon, LaserBullet)):
                                bullet.marked_for_deletion = True
                            if enemy.hp <= 0:
                                enemy.state = 'DEFEATED'
                                audio_manager.play('explosion_ground')
                    else:
                        if not isinstance(bullet, (WaveCannon, LaserBullet)):
                            bullet.marked_for_deletion = True
                        if not enemy.marked_for_deletion:
                            enemy.marked_for_deletion = True
                            rand_drop = random.random()
                            if rand_drop <= 0.10:
                                self.items.append(ShieldCapsule(enemy.x, enemy.y))
                            elif rand_drop <= 0.20:
                                self.items.append(WeaponCapsule(enemy.x, enemy.y))
                            elif rand_drop <= 0.30:
                                self.items.append(LaserCapsule(enemy.x, enemy.y))
                            if enemy.type == 'phaeon':
                                self.score += 500
                            else:
                                self.score += 100
                            audio_manager.play('explosion_air')
                    
            if isinstance(bullet, WaveCannon):
                for eb in self.enemy_bullets:
                    if self.is_colliding(bullet, eb):
                        eb.marked_for_deletion = True
                        self.score += 10

        for bomb in [b for b in self.bombs if b.exploded]:
            for enemy in [e for e in self.enemies if e.type == 'ground']:
                dx = bomb.x - enemy.x
                dy = bomb.y - enemy.y
                distance = (dx * dx + dy * dy) ** 0.5
                if distance < (20 + bomb.explosion_timer) + enemy.width / 2:
                    if not enemy.marked_for_deletion:
                        enemy.marked_for_deletion = True
                        rand_drop = random.random()
                        if rand_drop <= 0.10:
                            self.items.append(ShieldCapsule(enemy.x, enemy.y))
                        elif rand_drop <= 0.20:
                            self.items.append(WeaponCapsule(enemy.x, enemy.y))
                        elif rand_drop <= 0.30:
                            self.items.append(LaserCapsule(enemy.x, enemy.y))
                        self.score += 300

        if not self.game_over:
            for item in self.items:
                if self.is_colliding(self.player, item):
                    item.marked_for_deletion = True
                    if isinstance(item, ShieldCapsule):
                        self.player.shield_count = 5
                    elif isinstance(item, WeaponCapsule):
                        self.player.has_laser = False
                        self.player.weapon_level = min(3, getattr(self.player, 'weapon_level', 1) + 1)
                    elif isinstance(item, LaserCapsule):
                        self.player.has_laser = True

            for enemy in [e for e in self.enemies if e.type in ('air', 'phaeon', 'boss')]:
                if self.is_colliding(self.player, enemy):
                    if enemy.type != 'boss':
                        enemy.marked_for_deletion = True
                    self.hit_player()

            for eb in self.enemy_bullets:
                if self.is_colliding(self.player, eb):
                    eb.marked_for_deletion = True
                    self.hit_player()

    def is_colliding(self, rect1, rect2):
        return (rect1.x - rect1.width/2 < rect2.x + rect2.width/2 and
                rect1.x + rect1.width/2 > rect2.x - rect2.width/2 and
                rect1.y - rect1.height/2 < rect2.y + rect2.height/2 and
                rect1.y + rect1.height/2 > rect2.y - rect2.height/2)

    def hit_player(self):
        if getattr(self.player, 'shield_count', 0) > 0:
            self.player.shield_count -= 1
            audio_manager.play('player_hit')
            self.trigger_rumble(0.6, 0.6, 250)
        else:
            self.lose_life()

    def lose_life(self):
        self.lives -= 1
        audio_manager.play('player_hit')
        self.trigger_rumble(1.0, 1.0, 500)
        
        # Reset player charging state
        self.player.charge_timer = 0
        self.player.was_z_pressed = False
        self.player.played_complete = False
        self.player.played_complete2 = False
        self.player.shoot_diffusion_wave = False
        self.player.weapon_level = 1
        self.player.has_laser = False
        audio_manager.stop_charge()
        
        self.enemies = [e for e in self.enemies if e.y >= self.player.y - 100]
        self.enemy_bullets.clear()
        self.player.x = self.width / 2
        self.player.y = self.height - 100
        
        if self.lives <= 0:
            self.game_over = True
            audio_manager.play('game_over')

File: Entities/Game.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using XeviShot.Audio;
using XeviShot.Input;
using XeviShot.Settings;

namespace XeviShot.Entities;

/// <summary>
/// ゲーム全体のロジック、エンティティ管理、衝突判定を司るクラス
/// </summary>
public class Game
{
    public float Width { get; }
    public float Height { get; }

    public Player Player { get; }
    public Background Background { get; }

    public List<Bullet> Bullets { get; } = new();
    public List<LaserBullet> LaserBullets { get; } = new();
    public List<WaveCannon> WaveCannons { get; } = new();
    public List<Bomb> Bombs { get; } = new();
    public List<Enemy> Enemies { get; } = new();
    public List<EnemyBullet> EnemyBullets { get; } = new();
    public List<ItemCapsule> Items { get; } = new();

    public int Score { get; private set; }
    public int HighScore { get; private set; }
    public int Lives { get; private set; } = 3;
    public bool GameOver { get; private set; } = false;
    public bool StageClear { get; private set; } = false;
    public bool BossActive { get; private set; } = false;
    public bool BossSpawned { get; private set; } = false;

    private float _enemySpawnTimer = 0f;
    private float _enemySpawnInterval = 60f;
    private float _groundSpawnTimer = 0f;
    private const float GroundSpawnInterval = 180f;

    private int _frameCount = 0;
    private float _phaeonSpawnTimer = 0f;
    private const float PhaeonSpawnInterval = 300f;

    private static readonly Random Rand = new();

    public Game(float width, float height, int initialScore = 0)
    {
        Width = width;
        Height = height;
        Score = initialScore;
        HighScore = Math.Max(Score, SettingsManager.Current.HighScore);

        Player = new Player(width / 2f, height - 100f);
        Background = new Background(width, height);
    }

    public void Update(InputManager input)
    {
        if (GameOver || StageClear) return;

        // デバッグ用: 1キーで街(1分)、2キーで敵要塞(2分50秒)、Tキーでボス直前(2分55秒)へスキップ
        if (input.CityTestKey && _frameCount < 3600)
        {
            _frameCount = 3600;
        }
        else if (input.FortressTestKey && _frameCount < 10200)
        {
            _frameCount = 10200;
        }
        else if (input.BossTestKey && _frameCount < 10500)
        {
            _frameCount = 10500;
        }

        _frameCount++;

        // 3分経過（10800フレーム）でボス出現
        if (_frameCount >= 10800 && !BossActive && !BossSpawned)
        {
            Enemies.Add(new Boss(Width / 2f, -50f));
            BossActive = true;
            BossSpawned = true;
            AudioManager.Instance.PlayBossBgm();
        }

        // ボス撃破確認
        if (BossActive)
        {
            bool hasBoss = false;
            foreach (var e in Enemies)
            {
                if (e is Boss)
                {
                    hasBoss = true;
                    break;
                }
            }

            if (!hasBoss)
            {
                BossActive = false;
                StageClear = true;
                Enemies.Clear();
                EnemyBullets.Clear();
                if (HighScore > SettingsManager.Current.HighScore)
                {
                    SettingsManager.Current.HighScore = HighScore;
                    SettingsManager.Save();
                }
            }
        }

        Background.Update(_frameCount);
        Player.UpdateInput(input, Width, Height);

        // 自機の振動トリガー処理
        if (Player.RumbleTrigger != null)
        {
            string trigger = Player.RumbleTrigger;
            Player.RumbleTrigger = null;
            if (trigger == "charge_complete")
            {
                input.TriggerRumble(0.0f, 0.6f, 120);
            }
            else if (trigger == "charge_complete2")
            {
                input.TriggerRumble(0.0f, 1.0f, 200);
            }
        }

        // 1. 対空弾通常発射
        if (Player.ShootNormal)
        {
            Player.ShootNormal = false;
            if (Player.HasLaser)
            {
                LaserBullets.Add(new LaserBullet(Player.X, Player.Y - Player.Height / 2f, level: Player.WeaponLevel));
                Player.CooldownAir = 12;
            }
            else
            {
                int level = Player.WeaponLevel;
                if (level >= 3)
                {
                    float[] angles = { -25f, -8f, 8f, 25f };
                    float[] offsets = { -15f, -5f, 5f, 15f };
                    for (int i = 0; i < angles.Length; i++)
                    {
                        double rad = angles[i] * Math.PI / 180.0;
                        float vx = (float)(10.0 * Math.Sin(rad));
                        float vy = (float)(-10.0 * Math.Cos(rad));
                        Bullets.Add(new Bullet(Player.X + offsets[i], Player.Y - Player.Height / 2f, vx, vy));
                    }
                }
                else if (level == 2)
                {
                    Bullets.Add(new Bullet(Player.X - 10f, Player.Y - Player.Height / 2f));
                    Bullets.Add(new Bullet(Player.X + 10f, Player.Y - Player.Height / 2f));
                }
                else
                {
                    Bullets.Add(new Bullet(Player.X, Player.Y - Player.Height / 2f));
                }
                Player.CooldownAir = 10;
            }
            AudioManager.Instance.Play("laser");
        }

        // 2. 波動砲発射 (通常チャージ)
        if (Player.ShootWave)
        {
            Player.ShootWave = false;
            WaveCannons.Add(new WaveCannon(Player.X, Player.Y - Player.Height / 2f));
            AudioManager.Instance.Play("wave_cannon");
            input.TriggerRumble(0.6f, 0.4f, 250);
        }

        // 3. 拡散波動砲発射 (最大チャージ)
        if (Player.ShootDiffusionWave)
        {
            Player.ShootDiffusionWave = false;
            float[] angles = { -30f, -15f, 0f, 15f, 30f };
            const float baseSpeed = 12f;
            foreach (float angle in angles)
            {
                double rad = angle * Math.PI / 180.0;
                float vx = (float)(baseSpeed * Math.Sin(rad));
                float vy = (float)(-baseSpeed * Math.Cos(rad));
                WaveCannons.Add(new WaveCannon(Player.X, Player.Y - Player.Height / 2f, vx, vy));
            }
            AudioManager.Instance.Play("wave_cannon");
            input.TriggerRumble(0.9f, 0.7f, 450);
        }

        // 4. 対地ボム発射 (X または C または Pad B)
        if (input.FireGround || input.FireBoth)
        {
            if (Player.CooldownGround <= 0)
            {
                float targetX = Player.X;
                float targetY = Player.Y - Player.ReticleDistance;
                Bombs.Add(new Bomb(Player.X, Player.Y, targetX, targetY));
                Player.CooldownGround = 60;
                AudioManager.Instance.Play("bomb_launch");
            }
        }

        // 5. 敵の出現管理（ボス未出現時）
        if (!BossSpawned)
        {
            // 空中通常敵 (Zoldas)
            _enemySpawnTimer++;
            if (_enemySpawnTimer > _enemySpawnInterval)
            {
                SpawnEnemy("air");
                _enemySpawnTimer = 0;
                if (_enemySpawnInterval > 20f)
                {
                    _enemySpawnInterval -= 0.5f;
                }
            }

            // 地上敵砲台
            _groundSpawnTimer++;
            if (_groundSpawnTimer > GroundSpawnInterval)
            {
                SpawnEnemy("ground");
                _groundSpawnTimer = 0;
            }

            // 1分経過（3600フレーム）後から出現する高速強敵 (Phaeon)
            if (_frameCount > 3600)
            {
                _phaeonSpawnTimer++;
                if (_phaeonSpawnTimer > PhaeonSpawnInterval)
                {
                    _phaeonSpawnTimer = 0;
                    int numPhaeons = Rand.Next(1, 4);
                    float baseX = (float)(Rand.NextDouble() * (Width - 160) + 80);
                    for (int i = 0; i < numPhaeons; i++)
                    {
                        float spawnX = Math.Clamp(baseX + (i - (numPhaeons - 1) / 2f) * 40f, 30f, Width - 30f);
                        float spawnY = -50f - i * 30f;
                        Enemies.Add(new Enemy(spawnX, spawnY, "phaeon"));
                    }
                }
            }
        }

        // 6. 各エンティティの更新
        foreach (var b in Bullets) b.Update();
        foreach (var lb in LaserBullets) lb.Update();
        foreach (var wc in WaveCannons) wc.Update();

        foreach (var bomb in Bombs)
        {
            bool wasExploded = bomb.Exploded;
            bomb.Update();
            if (bomb.Exploded && !wasExploded)
            {
                input.TriggerRumble(0.4f, 0.2f, 150);
            }
        }

        foreach (var enemy in Enemies)
        {
            enemy.Update();
            if (enemy.ShootNow)
            {
                enemy.ShootNow = false;
                if (enemy.Type == "phaeon")
                {
                    float dx = Player.X - enemy.X;
                    float dy = Player.Y - enemy.Y;
                    float dist = (float)Math.Sqrt(dx * dx + dy * dy);
                    float vx = dist > 0 ? (dx / dist) * 6.0f : 0f;
                    float vy = dist > 0 ? (dy / dist) * 6.0f : 6.0f;
                    EnemyBullets.Add(new EnemyBullet(enemy.X, enemy.Y, vx, vy));
                }
                else if (enemy is Boss)
                {
                    float[] angles = { -40f, -20f, 0f, 20f, 40f };
                    const float bulletSpeed = 5.0f;
                    foreach (float angle in angles)
                    {
                        double rad = angle * Math.PI / 180.0;
                        float vx = (float)(bulletSpeed * Math.Sin(rad));
                        float vy = (float)(bulletSpeed * Math.Cos(rad));
                        EnemyBullets.Add(new EnemyBullet(enemy.X, enemy.Y + 20f, vx, vy));
                    }
                }
                else
                {
                    EnemyBullets.Add(new EnemyBullet(enemy.X, enemy.Y + enemy.Height / 2f));
                }
            }
        }

        foreach (var eb in EnemyBullets) eb.Update();
        foreach (var item in Items) item.Update();

        // 削除済みエンティティの一括除去
        Bullets.RemoveAll(b => b.MarkedForDeletion);
        LaserBullets.RemoveAll(lb => lb.MarkedForDeletion);
        WaveCannons.RemoveAll(wc => wc.MarkedForDeletion);
        EnemyBullets.RemoveAll(eb => eb.MarkedForDeletion);
        Bombs.RemoveAll(b => b.MarkedForDeletion);
        Enemies.RemoveAll(e => e.MarkedForDeletion);
        Items.RemoveAll(i => i.MarkedForDeletion);

        // 7. 衝突判定
        CheckCollisions(input);

        // 衝突により削除マークされたエンティティを除去
        Bullets.RemoveAll(b => b.MarkedForDeletion);
        LaserBullets.RemoveAll(lb => lb.MarkedForDeletion);
        WaveCannons.RemoveAll(wc => wc.MarkedForDeletion);
        EnemyBullets.RemoveAll(eb => eb.MarkedForDeletion);
        Bombs.RemoveAll(b => b.MarkedForDeletion);
        Enemies.RemoveAll(e => e.MarkedForDeletion);
        Items.RemoveAll(i => i.MarkedForDeletion);

        // ハイスコア更新チェック（現在のスコアが過去最高を超えたら更新）
        if (Score > HighScore)
        {
            HighScore = Score;
        }
    }

    private void SpawnEnemy(string type)
    {
        float x = (float)(Rand.NextDouble() * (Width - 60) + 30);
        float y = -50f;
        Enemies.Add(new Enemy(x, y, type));
    }

    private void CheckCollisions(InputManager input)
    {
        // 1. 自機通常弾 vs 空中敵 / ボス
        foreach (var bullet in Bullets)
        {
            if (bullet.MarkedForDeletion) continue;

            foreach (var enemy in Enemies)
            {
                if (enemy.Type is not ("air" or "phaeon") && enemy is not Boss) continue;
                if (enemy.MarkedForDeletion) continue;

                if (IsColliding(bullet, enemy))
                {
                    bullet.MarkedForDeletion = true;
                    HitEnemy(enemy);
                    break;
                }
            }
        }

        // 2. 自機レーザー弾 vs 空中敵 / ボス (貫通)
        foreach (var lb in LaserBullets)
        {
            if (lb.MarkedForDeletion) continue;

            foreach (var enemy in Enemies)
            {
                if (enemy.Type is not ("air" or "phaeon") && enemy is not Boss) continue;
                if (enemy.MarkedForDeletion) continue;

                if (IsColliding(lb, enemy))
                {
                    HitEnemy(enemy);
                }
            }
        }

        // 3. 波動砲 vs 空中敵 / ボス (貫通) & 敵弾かき消し
        foreach (var wc in WaveCannons)
        {
            if (wc.MarkedForDeletion) continue;

            // 敵への貫通攻撃
            foreach (var enemy in Enemies)
            {
                if (enemy.Type is not ("air" or "phaeon") && enemy is not Boss) continue;
                if (enemy.MarkedForDeletion) continue;

                if (IsColliding(wc, enemy))
                {
                    HitEnemy(enemy);
                }
            }

            // 敵弾のかき消し
            foreach (var eb in EnemyBullets)
            {
                if (eb.MarkedForDeletion) continue;

                if (IsColliding(wc, eb))
                {
                    eb.MarkedForDeletion = true;
                    Score += 10;
                }
            }
        }

        // 4. ボム爆発 vs 地上敵
        foreach (var bomb in Bombs)
        {
            if (!bomb.Exploded) continue;

            foreach (var enemy in Enemies)
            {
                if (enemy.Type != "ground" || enemy.MarkedForDeletion) continue;

                float dx = bomb.X - enemy.X;
                float dy = bomb.Y - enemy.Y;
                float dist = (float)Math.Sqrt(dx * dx + dy * dy);
                float radius = (20f + bomb.ExplosionTimer) + enemy.Width / 2f;

                if (dist < radius)
                {
                    enemy.MarkedForDeletion = true;
                    Score += 300;
                    DropItem(enemy.X, enemy.Y);
                }
            }
        }

        if (!GameOver)
        {
            // 5. 自機 vs アイテム
            foreach (var item in Items)
            {
                if (item.MarkedForDeletion) continue;

                if (IsColliding(Player, item))
                {
                    item.MarkedForDeletion = true;
                    if (item is ShieldCapsule)
                    {
                        Player.ShieldCount = 5;
                    }
                    else if (item is WeaponCapsule)
                    {
                        if (Player.HasLaser)
                        {
                            // レーザーモード → 通常弾カプセル: レーザー解除して WEAPON LEVEL 1 にリセット
                            Player.HasLaser = false;
                            Player.WeaponLevel = 1;
                        }
                        else
                        {
                            // 通常弾モード → 通常弾カプセル: レベルアップ (最大3)
                            Player.WeaponLevel = Math.Min(3, Player.WeaponLevel + 1);
                        }
                    }
                    else if (item is LaserCapsule)
                    {
                        if (!Player.HasLaser)
                        {
                            // 通常弾モード → レーザーカプセル: レーザー有効化して WEAPON LEVEL 1 に初期化
                            Player.HasLaser = true;
                            Player.WeaponLevel = 1;
                        }
                        else
                        {
                            // レーザーモード → レーザーカプセル: レベルアップ (最大3)
                            Player.WeaponLevel = Math.Min(3, Player.WeaponLevel + 1);
                        }
                    }
                }
            }

            // 6. 自機 vs 空中敵 / ボス
            foreach (var enemy in Enemies)
            {
                if (enemy.Type is not ("air" or "phaeon") && enemy is not Boss) continue;
                if (enemy.MarkedForDeletion) continue;

                if (IsColliding(Player, enemy))
                {
                    if (enemy is not Boss)
                    {
                        enemy.MarkedForDeletion = true;
                    }
                    HitPlayer(input);
                    break;
                }
            }

            // 7. 自機 vs 敵弾
            foreach (var eb in EnemyBullets)
            {
                if (eb.MarkedForDeletion) continue;

                if (IsColliding(Player, eb))
                {
                    eb.MarkedForDeletion = true;
                    HitPlayer(input);
                    break;
                }
            }
        }
    }

    private void HitEnemy(Entity entity)
    {
        if (entity is Boss boss)
        {
            if (boss.State is "ENTER" or "HOVER")
            {
                boss.Hp--;
                boss.FlashTimer = 5;
                if (boss.Hp <= 0)
                {
                    boss.State = "DEFEATED";
                    AudioManager.Instance.Play("explosion_ground");
                }
            }
        }
        else if (entity is Enemy enemy)
        {
            enemy.MarkedForDeletion = true;
            DropItem(enemy.X, enemy.Y);

            if (enemy.Type == "phaeon")
            {
                Score += 500;
            }
            else
            {
                Score += 100;
            }
            AudioManager.Instance.Play("explosion_air");
        }
    }

    private void DropItem(float x, float y)
    {
        double roll = Rand.NextDouble();
        if (roll <= 0.10)
        {
            Items.Add(new ShieldCapsule(x, y));
        }
        else if (roll <= 0.20)
        {
            Items.Add(new WeaponCapsule(x, y));
        }
        else if (roll <= 0.30)
        {
            Items.Add(new LaserCapsule(x, y));
        }
    }

    private void HitPlayer(InputManager input)
    {
        if (Player.ShieldCount > 0)
        {
            Player.ShieldCount--;
            AudioManager.Instance.Play("player_hit");
            input.TriggerRumble(0.6f, 0.6f, 250);
        }
        else
        {
            LoseLife(input);
        }
    }

    private void LoseLife(InputManager input)
    {
        Lives--;
        AudioManager.Instance.Play("player_hit");
        input.TriggerRumble(1.0f, 1.0f, 500);

        // 状態リセット
        Player.ChargeTimer = 0;
        Player.WasFireAirPressed = false;
        Player.PlayedComplete = false;
        Player.PlayedComplete2 = false;
        Player.ShootDiffusionWave = false;
        Player.WeaponLevel = 1;
        Player.HasLaser = false;
        AudioManager.Instance.StopCharge();

        // 画面内の敵弾と至近の敵をクリア
        Enemies.RemoveAll(e => e.Y < Player.Y - 100f);
        EnemyBullets.Clear();

        Player.X = Width / 2f;
        Player.Y = Height - 100f;

        if (Lives <= 0)
        {
            GameOver = true;
            if (HighScore > SettingsManager.Current.HighScore)
            {
                SettingsManager.Current.HighScore = HighScore;
                SettingsManager.Save();
            }
            AudioManager.Instance.Play("game_over");
        }
    }

    private static bool IsColliding(Entity r1, Entity r2)
    {
        return r1.X - r1.Width / 2f < r2.X + r2.Width / 2f &&
               r1.X + r1.Width / 2f > r2.X - r2.Width / 2f &&
               r1.Y - r1.Height / 2f < r2.Y + r2.Height / 2f &&
               r1.Y + r1.Height / 2f > r2.Y - r2.Height / 2f;
    }

    public void Draw(Graphics g)
    {
        // 1. 背景
        Background.Draw(g);

        // 2. 地上敵
        foreach (var e in Enemies)
        {
            if (e.Type == "ground") e.Draw(g);
        }

        // 3. ボム
        foreach (var b in Bombs) b.Draw(g);

        // 4. アイテムカプセル
        foreach (var item in Items) item.Draw(g);

        // 5. 自機
        if (!GameOver)
        {
            Player.Draw(g);
        }

        // 6. 空中敵・ボス
        foreach (var e in Enemies)
        {
            if (e.Type is "air" or "phaeon" || e is Boss) e.Draw(g);
        }

        // 7. 自機弾・敵弾
        foreach (var b in Bullets) b.Draw(g);
        foreach (var lb in LaserBullets) lb.Draw(g);
        foreach (var wc in WaveCannons) wc.Draw(g);
        foreach (var eb in EnemyBullets) eb.Draw(g);
    }
}
